src/Interpreters/GraceHashJoin.cpp

```diff
@@ -471,17 +472,24 @@ bool GraceHashJoin::alwaysReturnsEmptySet() const
     return hash_join_is_empty;
 }
 
-IBlocksStreamPtr GraceHashJoin::getNonJoinedBlocks(const Block &, const Block &, UInt64) const
+// There is a finished counter in JoiningTransform/DelayedJoinedBlocksWorkerTransform,
+// only the last processor could call this function to ensure all used flags have been inserted.
+// To support delayed stream mode, need to keep the hash join before next getDelayedBlocks call.
+IBlocksStreamPtr
+GraceHashJoin::getNonJoinedBlocks(const Block & left_sample_block_, const Block & result_sample_block_, UInt64 max_block_size_) const
 {
-    /// We do no support returning non joined blocks here.
-    /// TODO: They _should_ be reported by getDelayedBlocks instead
-    return nullptr;
+    return hash_join->getNonJoinedBlocks(left_sample_block_, result_sample_block_, max_block_size_);
```

Comment:
Sorry, the comment may be ambiguous. Here I mean the latest bucket. Eache bucket are handle by following steps
1. join left side with the hash table
2. read non-joined rows from the hash table

Reply:
@lgbo-ustc so, because scatter is also performed during loading, the resulting hash_join will have only corresponding rows according to the current number of buckets, is it correct?